Signal/src/view controllers/MessageComposeTableViewController.m

```diff
@@ -407,17 +476,36 @@ - (void)messageComposeViewController:(MFMessageComposeViewController *)controlle
 #pragma mark - Table View Data Source
 
 - (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
-    return 2;
+    return MessageComposeTableViewControllerSectionContacts + 1;
 }
 
 - (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
-    if (section == MessageComposeTableViewControllerSectionInvite) {
-        if (floor(NSFoundationVersionNumber) < NSFoundationVersionNumber_iOS_9_0) {
-            // Invite flow not supported on iOS8
-            return 0;
-        }
-        return 1;
+    // This logic will determine which one (if any) of the following special controls
+    // should be shown.  No more than one should be shown at a time.
+    BOOL showNonContactConversation = NO;
+    BOOL showInviteViaSMS = NO;
+    BOOL showInviteFlow = NO;
+
+    BOOL hasPhoneNumber = self.searchPhoneNumber.length > 0;
+    BOOL isKnownSignalUser = hasPhoneNumber && [self.phoneNumberAccountSet containsObject:self.searchPhoneNumber];
+    BOOL isInviteFlowSupported = floor(NSFoundationVersionNumber) >= NSFoundationVersionNumber_iOS_9_0;
```

Comment:
Yeah, just another example me trying to refactor this code as little as possible.  It's hard because I want to rewrite so much - I'm trying to restrict myself to changes that deliver high benefit.